reasoner/planner/ReasonerPlanner.java

```diff
@@ -50,34 +50,44 @@ public ReasonerPlanner(TraversalEngine traversalEng, ConceptManager conceptMgr,
     }
 
     public static ReasonerPlanner create(TraversalEngine traversalEng, ConceptManager conceptMgr, LogicManager logicMgr) {
-        return new GreedyCostSearch.OldPlannerEmulator(traversalEng, conceptMgr, logicMgr);
+        return new GreedyCostSearch.HeuristicPlannerEmulator(traversalEng, conceptMgr, logicMgr);
     }
 
-    public Plan plan(ResolvableConjunction conjunction, Set<Identifier.Variable.Retrievable> bounds) {
-        return planCache.get(new Pair<>(conjunction, bounds), cbPair -> computeResolvableOrdering(logicMgr.compile(cbPair.first()), cbPair.second()));
+    public Plan plan(ResolvableConjunction conjunction, Set<Variable> bounds) {
+        Pair<ResolvableConjunction, Set<Variable>> planKey = new Pair<>(conjunction, bounds);
+        Plan plan = planCache.getIfPresent(planKey);
+        if (plan == null) {
+            plan = computePlan(planKey.first(), planKey.second());
+            planCache.put(planKey, plan);
+        }
+        return plan;
+        // TODO: change back to below if we can fix the recursive update
+        //  return planCache.get(new Pair<>(conjunction, bounds), cbPair -> computePlan(cbPair.first(), cbPair.second()));
     }
 
-    abstract Plan computeResolvableOrdering(Set<Resolvable<?>> resolvables, Set<Identifier.Variable.Retrievable> bounds);
+    abstract Plan computePlan(ResolvableConjunction conjunction, Set<Variable> bounds);
 
-    boolean dependenciesSatisfied(Resolvable<?> resolvable, Set<Identifier.Variable.Retrievable> bounds, Map<Resolvable<?>, Set<Identifier.Variable.Retrievable>> dependencies) {
+    static boolean dependenciesSatisfied(Resolvable<?> resolvable, Set<Variable> bounds, Map<Resolvable<?>, Set<Variable>> dependencies) {
         return bounds.containsAll(dependencies.get(resolvable));
     }
 
     /**
      * Determine the resolvables that are dependent upon the generation of each variable
      */
-    static Map<Resolvable<?>, Set<Identifier.Variable.Retrievable>> dependencies(Set<Resolvable<?>> resolvables) {
-        Map<Resolvable<?>, Set<Identifier.Variable.Retrievable>> deps = new HashMap<>();
-        Set<Identifier.Variable.Retrievable> generated = iterate(resolvables).map(Resolvable::generating).filter(Optional::isPresent)
-                .map(Optional::get).map(ThingVariable::id).toSet();
-
-        Map<Identifier.Variable.Retrievable, Integer> unnegatedRefCount = new HashMap<>();
+    static Map<Resolvable<?>, Set<Variable>> dependencies(Set<Resolvable<?>> resolvables) {
+        // TODO: There may not be any generated->used dependencies since every use either triggers the rule or is unsatisfiable
+        Map<Resolvable<?>, Set<Variable>> deps = new HashMap<>();
```

Comment:
This is a todo for later. Relaxing all generated->used dependencies does causes tests to fail